design-decisions/2018-10-12-Condition-Split.md

```diff
@@ -0,0 +1,96 @@
+Splitting configurations upon new constraints on symbolic variables
+===================================================================
+
+Background
+----------
+
+When applying an axiom on a symbolic configuration new constraints may be
+introduced on symbolic variables, coming from side conditions, or directly
+from the unification process.
+
+### Example (from side conditions)
+
+Consider rewriting term `3 / A:Int` using
+
+`rule X:Int / Y:Int => X /Int Y requires Y =/=Int 0`
+which would be written as
+`Y =/= 0 = true /\ X:Int / Y:Int => X /Int Y`
+
+Then `3 / A:Int /\ (Y =/=Int 0 = true /\ X:Int / Y:Int)` will simplify to
+```
+3 / A:Int /\ (X:Int = 3 /\ Y:Int = A:Int)
+          /\ (A:Int =/=Int 0) = true
+```
+which is satisfiable, but not `T`.
+
+### Example (from unification)
+
+Consider rewriting term `if A:Int <Int 0 then -1 else 1` using
+
+`rule if true then E1:Exp else E2:Exp => E1`
+
+Then `if A:Int <Int 0 then -1 else 1 /\ if true then E1:Exp else E2:Exp` will
+simplify to
+```
+if A:Int <Int 0 then -1 else 1 /\ (E1:Exp = -1 /\ E2:Exp = 1)
+                               /\ A:Int <Int 0 = true
+```
+which is satisfiable, but not `T`.
+
+Problem
+-------
+
+Simply continuing the rewriting process when a new constraint is introduced
+(on some symbolic variable) is not sound when the constraint does not simplify
+to top.  The reason is that, going further, the coverage is reduced to the
+inputs which satisfy the constraint.
+
+Questions
+---------
+
+* When/how do we decide that a new constraint is satisfiable but not top?
+
+* How about such constraints coming from function application?
+
+Decision: do not apply function axioms if they introduce new constraints
+------------------------------------------------------------------------
+
+We might revert this decision later, but for now this is a safe bet, as it
+prevents undesirable unrolling of functions.
+
+Decision: check whether new constraints are implied by existing ones
+--------------------------------------------------------------------
+
+To exhibit the problem described above, a new constraint must not be implied
+by the existing constraint, and their conjunction should be satisfiable but
```

Comment:
Isn't talking about the conjunction being top totally redundant? `CC /\ NC` can only be top if `CC -> NC` was true and `CC` was already `top, right?